### PR TITLE
Adds basic crate-level documentation

### PR DESCRIPTION
**Pull request by zslayton**
Provides examples for creating, reading, and writing `Element`s.

image

image

image

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/binary/non_blocking/raw_binary_reader.rs
Comment (zslayton) [inline src/binary/non_blocking/raw_binary_reader.rs:1]: 🗺️ `cargo doc` pointed out some nits in this file and the next.

File: src/element/owned.rs [hunk 1]
```diff
@@ -598,6 +598,18 @@ impl From<Symbol> for Value {
...
+    fn from(value: &[u8]) -> Self {
+        Value::Blob(value.into())
+    }
+}
+
+impl From<Vec<u8>> for Value {
+    fn from(value: Vec<u8>) -> Self {
+        Value::Blob(value)
+    }
+}
```
Comment (zslayton) [inline src/element/owned.rs:611]: 🗺️ While writing the doc tests, I realized we didn't have an `Into<Value>` impl for the `blob`-ish Rust types.

File: src/lib.rs [hunk 2]
```diff
@@ -29,8 +187,11 @@ mod symbol_table;
 mod system_reader;
 mod writer;
 
+#[doc(inline)]
 pub use data_source::IonDataSource;
+#[doc(inline)]
 pub use raw_symbol_token::RawSymbolToken;
+#[doc(inline)]
```
Comment (zslayton) [inline src/lib.rs:194]: 🗺️ `#[doc(inline)]` prevents these from showing up in a big "Re-exports" list in the docs. We'll probably add this to more types before 1.0.

**Review by zslayton — COMMENTED** [review]
🗺️ PR tour

File: src/lib.rs [hunk 3]
```diff
@@ -1,6 +1,164 @@
...
+//! writer.flush()?;
+//! assert_eq!(
+//!   "{foo: \"hello\", bar: true, baz: [4, 5, 6]}".as_bytes(),
+//!   writer.output().as_slice()
+//! );
+//! # Ok(())
+//! # }
+//! ```
 
+use element::owned::Element;
```
Comment (desaikd) [inline src/lib.rs:161]: clippy suggests this unused import

File: src/lib.rs [hunk 4]
```diff
@@ -1,6 +1,164 @@
...
+//! # fn main() -> IonResult<()> {
+//! use std::fs::File;
+//! use ion_rs::element::owned::Element;
+//! use ion_rs::element::reader::ElementReader;
+//! use ion_rs::ReaderBuilder;
+//! let ion_file = File::open("/foo/bar/baz.ion").unwrap();
+//! let mut reader = ReaderBuilder::default().build(ion_file)?;
+//! // A simple pretty-printer
+//! for element in reader.elements() {
+//!     println!("{}", element?)
```
Comment (desaikd) [inline src/lib.rs:102]: Should we add a sample file content and output for this example?

**Review by desaikd — APPROVED** [review]
Few minor comments otherwise looks good!
Really nice to have those macro examples 🚀

File: src/lib.rs (hunk 4 above)
Comment (zslayton) [reply src/lib.rs:102]: It does stand out as the only example without an `assert_eq!`, but I figured seeing how to access the elements in a file would be enough for folks to use the other tests' examples. I'm going to leave this as-is for the moment to get v0.16.0 out, but I'm open to changing it in the future.

File: src/lib.rs (hunk 3 above)
Comment (zslayton) [reply src/lib.rs:161]: Fixed!

**Pull request merged**